### PR TITLE
Update azure_rm_managed_disk.py --add zones

### LINKED ISSUE #51321: Availability Zones for managed disks
##### SUMMARY
Implement Availability Zones for managed disks

##### ISSUE TYPE
Please add Availibility Zones support for  managed disks (azure_rm_managed_disk),  similar to "az disk create --name xxx --zone X ..."

##### COMPONENT NAME
azure_rm_managed_disk

##### ADDITIONAL INFORMATION
related to https://github.com/ansible/ansible/pull/49243

```yaml

```

### PR DESCRIPTION
**Pull request by Fred-sun**
Implement Availability Zones for managed disk (fixes #51321)

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

azure_rm_managed_disk.py
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 1]
```diff
@@ -220,6 +224,9 @@ def __init__(self):
             ),
             managed_by=dict(
                 type='str'
+            ),
+            zones=dict(
+                type='list'
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: type='list', elements='str'

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 2]
```diff
@@ -85,7 +85,10 @@
     tags:
         description:
             - Tags to assign to the managed disk.
-
+    zones:
+        description:
+            - A list of avaliablity Zones for you managed disk.
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: version_added: "2.8"

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 3]
```diff
@@ -258,6 +266,8 @@ def exec_module(self, **kwargs):
         resource_group = self.get_resource_group(self.resource_group)
         if not self.location:
             self.location = resource_group.location
+        # convert elements to ints
+        self.zones = [int(i) for i in self.zones] if self.zones else None
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: remove this line and add it to spec. BTW, the SDK accept the zones as `[str]`

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 4]
```diff
@@ -240,6 +247,7 @@ def __init__(self):
         self.os_type = None
         self.disk_size_gb = None
         self.tags = None
+        self.tags = None
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: redundant

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 5]
```diff
@@ -335,6 +345,7 @@ def generate_managed_disk_property(self):
         creation_data = {}
         disk_params['location'] = self.location
         disk_params['tags'] = self.tags
+        disk_params['zones'] = self.zones
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: Please also update the comments at line 343

**Review by yuwzho — CHANGES_REQUESTED** [review]
The PR looks good 👍  Please also add a  test.

BTW, to refer the issue, use `fixes` instead of `fixed by`

**Comment by Fred-sun** [comment]
@yuwzho Thanks Yuwzho, I have updated it again, please help to check!

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 6]
```diff
@@ -100,6 +104,7 @@
         location: eastus
         resource_group: myResourceGroup
         disk_size_gb: 4
+        zones: 2
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: This sample doesn't match the doc

**Review by yuwzho — CHANGES_REQUESTED** [review]

**Comment by yuwzho** [comment]
Please also address the other 4 comments. BTW the integration test is under https://github.com/ansible/ansible/blob/devel/test/integration/targets/azure_rm_managed_disk/tasks/main.yml

**Comment by Fred-sun** [comment]
Thanks Yuwei, I submit the new test!

**Comment by Fred-sun** [comment]
ready_for_review

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py (hunk 1 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: add elements type

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py (hunk 5 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: remove the corresponding comments

File: test/integration/targets/azure_rm_managed_disk/tasks/main.yml [hunk 7]
```diff
@@ -62,6 +63,24 @@
        - output.state.id is defined
        - output.state.os_type == None
 
+ - name: Create new managed disk with zones
+   azure_rm_managed_disk:
+       resource_group: "{{ resource_group }}"
+       name: "md{{ rpfx }}3"
+       storage_account_type: "Standard_LRS"
+       disk_size_gb: 1
+       zones: 1
```
Comment (yuwzho) [inline test/integration/targets/azure_rm_managed_disk/tasks/main.yml]: this is not a list

```
zone:
   - 1
   - 2
```

**Review by yuwzho — CHANGES_REQUESTED** [review]
Please also add the sample

**Comment by yuwzho** [comment]
Please add ready_for_review tag after addressing all comments

**Comment by Fred-sun** [comment]
fixes #51321

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py (hunk 1 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: ?

**Review by yuwzho — CHANGES_REQUESTED** [review]
Please modify the  zones to `zones=dict(type='list', elements='int')` in arg_spec according to the comments

**Comment by Fred-sun** [comment]
@yuwzho Please help re-check this when you're available! Thanks!

**Review by yuwzho — APPROVED** [review]

**Comment by yuwzho** [comment]
shipit

**Comment by Fred-sun** [comment]
@yuwzho Could we can merge this PR? Thanks!

**Comment by jhawkesworth** [comment]
I was able to get this to work but it is strange that zones must be a list but also must only contain 1 entry
If you supply 'zones' like this:
```yaml
            zones: 
              - 1 
              - 2
              - 3 
```                
The module fails with this error
```shell
Error creating the managed disk: Azure Error: ResourceTypeNotSupportMultiZones
Message: The resource type 'disks' does not support multiple availability zones. The provided zones are '1,2,3'. Please provide a single availability zone.
```
If fixed it by supply a zone based on a number taken from hostname that the disk is for.
```yaml
        zones:
          - "{{ ((inventory_hostname[-1]|int + 2) % 3) + 1 }}"
```

So it think maybe zones should have 'string' type not 'list' type and maybe called 'zone' (singular) not 'zones' (plural).  Or maybe I don't understand?

**Comment by Fred-sun** [comment]
@jhawkesworth Thanks for your test, I will check!

**Comment by Fred-sun** [comment]
@yuwzho I changed the type of zone from list to string as suggested by the comment, but I'm not sure it makes sense. please help to review. Thanks!

**Comment by yuwzho** [comment]
I don't think we should make this field as a single string. My suggestion is to find out why this option can be a list of int, for some specific SKU or others.

**Comment by Fred-sun** [comment]
@yuwzho You mean a disk can be used in different zones? it can use list to set more than one zones! Thanks！

**Comment by Fred-sun** [comment]
When creating a disk, it can use multiple zones.
like this:
···e.g. 
            zones: 
              - 1 
              - 2
              - 3

**Comment by jhawkesworth** [comment]
I do not understand why when I tried this
```yaml
            zones: 
              - 1 
              - 2
              - 3 
```
 on 27 March 2019 for managed disks in US East 2 region I got the
```bash
Error creating the managed disk: Azure Error: ResourceTypeNotSupportMultiZones
Message: The resource type 'disks' does not support multiple availability zones. The provided zones are '1,2,3'. Please provide a single availability zone.
```
error.   

It sounds like zones needs to be a list but it is not yet supported for disks in US East 2 perhaps?

Is there a way to find out if a resource type supports multiple zones yet?  Its not great if user has to discover this by running playbook and watching for failure.

**Comment by Fred-sun** [comment]
@yuwzho Could you can reply the contributor's confuse? I don't know much about the zones! Thanks!

**Comment by Fred-sun** [comment]
@yuwzho update according by the comment, please review! Thanks!

**Comment by Fred-sun** [comment]
ready_for_review

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 8]
```diff
@@ -330,11 +345,16 @@ def _get_vm(self, name):
             self.fail("Error getting virtual machine {0} - {1}".format(name, str(exc)))
 
     def generate_managed_disk_property(self):
-        # TODO: Add support for EncryptionSettings, DiskIOPSReadWrite, DiskMBpsReadWrite, Zones
+        # TODO: Add support for EncryptionSettings, DiskIOPSReadWrite, DiskMBpsReadWrite
         disk_params = {}
         creation_data = {}
+        disk_params['zones'] = []
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: disk_params['zones'] = [self.zones] if self.zones and self.zones != 'None'

**Review by yuwzho — CHANGES_REQUESTED** [review]
Please also modify the `is_different` method to compare the zones' change

**Comment by Fred-sun** [comment]
When the zones on the disk are not matching zones in the playbook, it will get following error:
```e.g.

fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error creating the managed disk: Azure Error: BadRequest\nMessage: Availability zone must not be changed on existing resource 'testdisk'. Existing zone '1', new zone '2'.."}

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 9]
```diff
@@ -85,6 +85,15 @@
...
         description:
             - Tags to assign to the managed disk.
+    zones:
+        description:
+            - "Allowed values: 1, 2, 3, None."
+        choices:
+            - 1
+            - 2
+            - 3
+            - None
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: "None" here may be a bit confusing.
Choices are numeric here.... but strings afterwards

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 10]
```diff
@@ -385,6 +402,9 @@ def is_different(self, found_disk, new_disk):
         if new_disk.get('tags') is not None:
             if not found_disk['tags'] == new_disk['tags']:
                 resp = True
+        if new_disk.get('zones') is not None:
+            if not found_disk['zones'] == new_disk['zones']:
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: The found_disk['zones'] can be `str('')` while it the same as new_disk as None. I would change the `to_dict()` method as another comment.

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 11]
```diff
@@ -174,7 +183,8 @@ def managed_disk_to_dict(managed_disk):
         disk_size_gb=managed_disk.disk_size_gb,
         os_type=managed_disk.os_type.lower() if managed_disk.os_type else None,
         storage_account_type=managed_disk.sku.name if managed_disk.sku else None,
-        managed_by=managed_disk.managed_by
+        managed_by=managed_disk.managed_by,
+        zones=managed_disk.zones if managed_disk.zones else None
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: `zone=managed_disk.zones[0] if managed_disk.zones and len(managed_disk.zones) > 0 else ''`

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 12]
```diff
@@ -220,6 +230,10 @@ def __init__(self):
             ),
             managed_by=dict(
                 type='str'
+            ),
+            zones=dict(
+                type='str',
+                choices=['None', '1', '2', '3']
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: change the `None` to  `''`

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 13]
```diff
@@ -330,11 +345,13 @@ def _get_vm(self, name):
             self.fail("Error getting virtual machine {0} - {1}".format(name, str(exc)))
 
     def generate_managed_disk_property(self):
-        # TODO: Add support for EncryptionSettings, DiskIOPSReadWrite, DiskMBpsReadWrite, Zones
+        # TODO: Add support for EncryptionSettings, DiskIOPSReadWrite, DiskMBpsReadWrite
         disk_params = {}
         creation_data = {}
         disk_params['location'] = self.location
         disk_params['tags'] = self.tags
+        if self.zones and self.zones != "None":
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]:
```python
if self.zone:
    disk_params['zones'] = [self.zone]
```

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py (hunk 9 above)
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: `- ''`

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 14]
```diff
@@ -85,6 +85,15 @@
     tags:
         description:
             - Tags to assign to the managed disk.
+    zones:
+        description:
+            - "Allowed values: 1, 2, 3, None."
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: `"Allowed values: 1, 2, 3, ''."`

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 15]
```diff
@@ -240,6 +254,7 @@ def __init__(self):
         self.os_type = None
         self.disk_size_gb = None
         self.tags = None
+        self.zones = None
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: Change to zone

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 16]
```diff
@@ -85,6 +85,15 @@
     tags:
         description:
             - Tags to assign to the managed disk.
+    zones:
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: change to `zone`

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py (hunk 9 above)
Comment (Fred-sun) [reply lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: @zikalino How do you think this should be changed? changed like the following:
```e.g.
            - "Allowed values: 1, 2, 3."
        choices:
            - 1
            - 2
            - 3

File: lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py [hunk 17]
```diff
@@ -220,6 +230,10 @@ def __init__(self):
             ),
             managed_by=dict(
                 type='str'
+            ),
+            zones=dict(
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_managed_disk.py]: `zone=dict(`

**Review by yuwzho — CHANGES_REQUESTED** [review]
Reserve `zones`, Let's use `zone` currently to support only one zone.

**Comment by yungezz** [comment]
@Fred-sun  is this one ready for  review?

**Comment by Fred-sun** [comment]
@yungezz Yes, I forgot to add label!

**Comment by Fred-sun** [comment]
ready_for_review

**Comment by yuwzho** [comment]
shipit

Have you tested it? @Fred-sun

**Comment by Fred-sun** [comment]
@yuwzho Yes, I have tested it, and I will update the test result to comment later! Thanks!

**Pull request merged**